draft-ietf-httpbis-layered-cookies.md

```diff
@@ -804,30 +804,59 @@ agent MUST remove from the cookie store all cookies whose expiry-time is null.
 
 To **Remove Excess Cookies for a Host** given a host _host_:
 
-1. Let _insecureCookies_ be all cookies in the user agent's cookie store whose host is host-equal to _host_ and whose secure is false.
+1. Let _insecureCookies_ be a list of references to all cookies in the user agent's cookie store whose host is host-equal
+   to _host_ and whose secure is false.
 
 1. Sort _insecureCookies_ by earliest last-access-time first.
 
-1. Let _secureCookies_ be all cookies in the user agent's cookie store whose host is host-equal to _host_ and whose secure is true.
+1. Let _secureCookies_ be a list of references to all cookies in the user agent's cookie store whose host is host-equal
+   to _host_ and whose secure is true.
 
 1. Sort _secureCookies_ by earliest last-access-time first.
 
+1. Let _excessHostCookies_ be an empty list of cookies.
+
 1. While _insecureCookies_'s size + _secureCookies_'s size is greater than the user agent's total cookies-per-host limit:
 
-    1. If _insecureCookies_ is not empty, remove the first item of _insecureCookies_ and
-    delete the corresponding cookie from the user agent's cookie store.
+    1. If _insecureCookies_ is not empty:
+
+        1. Let _cookie_ be the first item of _insecureCookies_.
+
+        1. Remove _cookie_ from _insecureCookies_.
+
+        1. Remove _cookie_ from the user agent's cookie store.
+
+        1. Append _cookie_ to _excessHostCookies_.
+
+    1. Otherwise:
+
+        1. Let _cookie_ be the first item of _secureCookies_.
+
+        1. Remove _cookie_ from _secureCookies_.
 
-    1. Otherwise, remove the first item of _secureCookies_ and delete the corresponding cookie from the user agent's cookie store.
+        1. Remove _cookie_ from the user agent's cookie store.
+
+        1. Append _cookie_ to _excessHostCookies_.
```

Comment:
Shouldn't we return _excessHostCookies_ at this point? Do we know for a fact that cookie-related events are fired for excess-cookie removals?